File: Graph/1971. Find if Path Exists in Graph/Optimal.java
class Solution {
    public boolean validPath(int n, int[][] edges, int source, int destination) {
        boolean[] seen = new boolean[n];
        seen[source] = true;
        boolean thorough = false;
        while(!seen[destination] && !thorough){
            thorough = true;
            for(int[] edge : edges){
                if(seen[edge[0]] ^ seen[edge[1]]){
                    seen[edge[0]] = true;
                    seen[edge[1]] = true;
                    if(seen[destination]){return true;}
                    thorough = false;
                }
            }
        }
        return seen[destination];
    }
}

File: Graph/1971. Find if Path Exists in Graph/Optimal.py
class Solution:
    def validPath(self, n: int, edges: List[List[int]], source: int, destination: int) -> bool:
        seen = [False] * n
        seen[source] = True
        thorough = False
        while not (thorough or seen[destination]):
            thorough = True
            for u, v in edges:
                if seen[u] ^ seen[v]:
                    seen[u] = seen[v] = True
                    if seen[destination]: return True
                    thorough = False
        return seen[destination]